bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+// This example disables best-effort de-duplication, which allows for higher throughput.
+func (i *bigQueryMetric) Save() (map[string]bq.Value, string, error) {
+	ret := map[string]bq.Value{
+		"metricname": i.metricname,
+		"tags":       i.tags,
+		"value":      i.value,
+		"timestamp":  i.timestamp,
+	}
+	return ret, bq.NoDedupeID, nil
+}
+
+func getMetricsTableSchema() bq.Schema {
+	databaseSchema := bq.Schema{
+		{Name: "metricname", Type: bq.StringFieldType, Description: "metric name", Required: true},
+		{Name: "tags", Type: bq.StringFieldType, Description: "metric attribute tags"},
+		{Name: "value", Type: bq.FloatFieldType, Description: "metric value"},
+		{Name: "timestamp", Type: bq.TimestampFieldType, Description: "sample collection timestamp", Required: true},
+	}
+	return databaseSchema
+}
+
+// Produce a formatted string of the table identifier.
+func (t bigQueryTable) formatDatasetId() string {
+	return fmt.Sprintf("`%s.%s`", t.project, t.dataset)
+}
+
+// Produce a formatted string of the dataset identifier.
+func (t bigQueryTable) formatTableId() string {
+	return fmt.Sprintf("`%s.%s.%s`", t.project, t.dataset, t.tablename)
+}
+
+// Check if a BigQuery table exists based on reading its metadata.
+func (t bigQueryTable) isTableExist(session bigQuerySession) bool {
+	_, err := session.client.Dataset(t.dataset).Table(t.tablename).Metadata(session.ctx)
+	return err == nil
+}
+
+// Normalize BigQuery table references by filling in with defaults, as needed.
+func normalizeTableRef(t *bigQueryTable) {
+	// Always use the default project.
+	t.project = bigQueryTableDefault.project
+
+	// If either of the following were not found in the protobuf message,
+	// their zero value will show up here.
+	if len(t.dataset) == 0 {
+		t.dataset = bigQueryTableDefault.dataset
+	}
+	if len(t.tablename) == 0 {
+		t.tablename = bigQueryTableDefault.tablename
+	}
+}
+
+// Create a new dataset using an explicit destination location.
+func createDataset(w io.Writer, t bigQueryTable, location string, session bigQuerySession) error {
+	meta := &bq.DatasetMetadata{
+		Location: location, // See https://cloud.google.com/bigquery/docs/locations
+	}
+	if err := session.client.Dataset(t.dataset).Create(session.ctx, meta); err != nil {
+		return fmt.Errorf("Error creating dataset '%s': %s\n", t.formatDatasetId(), err)
+	}
+	fmt.Fprintf(w, "Created dataset %s\n", t.formatDatasetId())
+	return nil
+}
+
+// Create a BigQuery table with a predefined schema.
+func createTable(w io.Writer, t bigQueryTable, session bigQuerySession) error {
+	// Always attempt to create the dataset that holds the table.
+	// This is faster than first querying the table metadata,
+	// then creating the dataset if not present.
+	// We expect an "already exists" error if the dataset currently
+	// exists for the project, which is the normal case.
+	if err := createDataset(w, t, databaseLocation, session); err != nil {
+		if !strings.Contains(strings.ToLower(err.Error()), "already exist") {
```

Comment:
This should probably be something like:

```
var apiErr *googleapi.Error
if errors.As(err, &apiErr)  && apiErr.Code == /* whatever the HTTP code for already exists is */ {
  // ignore
} else {
  // some other error worth handling; add context and return
}
```

Reply:
I verified that this approach works, checking for http.StatusConflict (409 already exists). However, this is only needed when creating a dataset or table in BigQuery, which the endpoint no longer does, so the entire function containing this logic is being removed.